<launch>
  <arg name="ns" default="tracking_ros" />
  <arg name="model_type" default="vit_h" />
  <arg name="device" default="cuda:0" />
  <arg name="prompt_mode" default="true" />
  <arg name="num_slots" default="-1" />

  <arg name="compressed" default="false"/>
  <arg name="input_image" default="/kinect_head/rgb/image_rect_color"/>
  <arg name="input_point" default="/$(arg ns)/processed_image/screenpoint" />
  <arg name="input_bbox" default="/$(arg ns)/processed_image/screenrectangle" />
  <arg name="processed_image" default="/$(arg ns)/processed_image"/>
  <arg name="segmentation_image" default="/$(arg ns)/segmentation_image" />
  <arg name="prompt_image" default="/$(arg ns)/prompt_image" />
  <arg name="input_segmentation" default="/$(arg ns)/input_segmentation" />
  <arg name="output_segmentation" default="/$(arg ns)/segmentation" />


  <group ns="$(arg ns)">
    <!-- process image and repub -->
    <node if="$(arg compressed)" name="raw_image_transport" pkg="image_transport" type="republish" args="compressed">
      <remap from="in" to="$(arg input_image)"/>
      <remap from="out" to="$(arg processed_image)"/>
    </node>
    <node unless="$(arg compressed)" name="raw_image_transport" pkg="image_transport" type="republish" args="raw">
      <remap from="in" to="$(arg input_image)"/>
      <remap from="out" to="$(arg processed_image)"/>
    </node>

    <!-- sam node -->
    <include file="$(find tracking_ros)/launch/sam_node.launch" >
      <arg name="ns" value="$(arg ns)" />
      <arg name="model_type" value="$(arg model_type)" />
      <arg name="device" value="$(arg device)" />
      <arg name="prompt_mode" value="$(arg prompt_mode)" />
      <arg name="num_slots" value="$(arg num_slots)" />
      <arg name="input_point" value="$(arg input_point)" />
      <arg name="input_bbox" value="$(arg input_bbox)" />
      <arg name="input_image" value="$(arg processed_image)"/>
      <arg name="prompt_image" value="$(arg prompt_image)"/>
      <arg name="input_segmentation" value="$(arg input_segmentation)" />
    </include>

    <!-- tracking node -->
    <include file="$(find tracking_ros)/launch/track_node.launch" >
      <arg name="ns" value="$(arg ns)" />
      <arg name="device" value="$(arg device)" />
      <arg name="input_image" value="$(arg processed_image)"/>
      <arg name="input_segmentation" value="$(arg input_segmentation)" />
      <arg name="output_image" value="$(arg segmentation_image)"/>
      <arg name="output_segmentation" value="$(arg output_segmentation)" />
    </include>

    <!-- interactive prompter -->
    <node if="$(arg prompt_mode)" name="image_view2" pkg="image_view2" type="image_view2" output="screen" >
      <remap from="image" to="$(arg processed_image)" />
      <param name="use_window" value="false" />
    </node>
    <node if="$(arg prompt_mode)" pkg="jsk_rqt_plugins" type="rqt_image_view2" name="prompter"
          output="screen">
      <remap from="event" to="image_view2/event" />
      <remap from="image_marked" to="image_marked" />
    </node>
    <node if="$(arg prompt_mode)" name="prompter_button"
          pkg="jsk_rqt_plugins" type="rqt_tabbed_buttons" output="screen" clear_params="true">
      <rosparam command="load" file="$(find tracking_ros)/resource/remote_tabbed_layout.yaml" />
    </node>

    <!-- visualize segmentation -->
    <node if="$(arg prompt_mode)" name="image_view" pkg="image_view" type="image_view" respawn="false" output="screen">
    <remap from="image" to="$(arg prompt_image)" />
    </node>

  </group>
</launch>
